### PR TITLE
Add guice7 (jakarta.inject) module

### LINKED ISSUE #206: guice 7 / jakarta inject support?
Trying to migrate an app using the guice module to guice 7 which uses jakarta.inject package/namespace, and seeing this error:

```
java.lang.NoClassDefFoundError: javax/inject/Inject
	at com.fasterxml.jackson.module.guice.GuiceAnnotationIntrospector._findGuiceInjectId(GuiceAnnotationIntrospector.java:44)
	at com.fasterxml.jackson.module.guice.GuiceAnnotationIntrospector.findInjectableValue(GuiceAnnotationIntrospector.java:18)
	at com.fasterxml.jackson.databind.introspect.AnnotationIntrospectorPair.findInjectableValue(AnnotationIntrospectorPair.java:309)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector._addInjectables(POJOPropertiesCollector.java:902)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.collectAll(POJOPropertiesCollector.java:464)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.getJsonValueAccessor(POJOPropertiesCollector.java:286)
	at com.fasterxml.jackson.databind.introspect.BasicBeanDescription.findJsonValueAccessor(BasicBeanDescription.java:258)
	at com.fasterxml.jackson.databind.ser.BasicSerializerFactory.findSerializerByAnnotations(BasicSerializerFactory.java:393)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:225)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:174)
	at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1507)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1455)
	at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:556)
	at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:834)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:307)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4719)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3964)
```

Assuming something would need to be done to support javax.inject.Inject and/or jakarta.inject.Inject based on runtime dependency availability?  seems like a mess but is there an existing workaround I'm missing?

### PR DESCRIPTION
**Pull request by josephlbarnett**
Copy guice (javax.inject) module over to a new guice 7 (jakarta.inject) module.  Replace javax imports with jakarta imports, and leave everything else the same.

Fixes #206

### CODE REVIEW

**Comment by josephlbarnett** [comment]
(will rework this based on 3 vs/ 2.16 package changes....)

**Review by GedMarc — APPROVED** [review]

File: guice7/src/moditect/module-info.java
Comment (GedMarc) [inline guice7/src/moditect/module-info.java:5]: requires transitive

File: guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java
Comment (GedMarc) [inline guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java:103]: can this be split out, to allow overriding on the bindings, or to rescope to object mapper

**Comment by cowtowncoder** [comment]
Sounds good, happy to merge (we have CLA).

Just one question: I guess "guice7" works (and I think it's best to leave existing id of "guide" etc as-is for compatibility), but I wonder if alternative of "guice-jakarta" would make sense wrt later Guice versions? (since this would work for Guice 8.x etc).

I don't have strong opinion here but many other modules add "jakarta".

**Comment by josephlbarnett** [comment]
I dont have a strong opinion either way -- I did notice the latest guice main branch commit removes java8 support, so not sure if that changes anything about how you'd think about these guice modules?

**Comment by cowtowncoder** [comment]
> I dont have a strong opinion either way -- I did notice the latest guice main branch commit removes java8 support, so not sure if that changes anything about how you'd think about these guice modules?

Hmmh. That gets more difficult wrt keeping Guice module here while building others for Java 8 since ideally we wouldn't use later JDKs for building (just to avoid accidental beyond-JDK-8 deps for modules not meant to require later baseline).
Some Jackson repos do require JDK 11 or later, but in those cases they are less problematic (JAX-RS/Jakarta-RS provider; Hibernate).

So let's go with `guice7` name that's fine. I'll try to get this merged soon.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
@josephlbarnett One question before merging -- have you addressed @GedMarc's concerns? If not, if you could do that (make change if they make sense which probably does for module-info -- or add a note why change doesn't make sense).

File: guice7/src/moditect/module-info.java
Comment (josephlbarnett) [reply guice7/src/moditect/module-info.java:5]: transitive on  com.fasterxml.jackson.annotation?  ok, i'll defer to you on module-info stuff but want to be sure I get it right.  should that change also be made to the existing guice module?

File: guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java
Comment (josephlbarnett) [reply guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java:103]: possibly, but not sure exactly what that would look like.  feels like a separate PR against existing guice module (and this new one if timing of that PR is after this gets in) would make more sense to me?

File: guice7/src/moditect/module-info.java
Comment (cowtowncoder) [reply guice7/src/moditect/module-info.java:5]: If change was made here, yes, I think similarly to existing. I don't have strong opinion.

File: guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java
Comment (cowtowncoder) [reply guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java:103]: I think that (possible separate follow-up PR) makes sense assuming this is not a deviation from existing Guice module.

Comment (GedMarc) [reply guice7/src/main/java/com/fasterxml/jackson/module/guice7/ObjectMapperModule.java:103]: Yeah agreed - will get started with that (y)

**Review by GedMarc — APPROVED** [review]

**Review by GedMarc — APPROVED** [review]

File: guice7/src/main/java/com/fasterxml/jackson/module/guice7/GuiceAnnotationIntrospector.java
Comment (cowtowncoder) [inline guice7/src/main/java/com/fasterxml/jackson/module/guice7/GuiceAnnotationIntrospector.java:1]: Whops. guice -> guice7; will fix.

**Comment by cowtowncoder** [comment]
Merged; also converted `master` to `tool.jackson` (2.x -> 3.x conversion)

File: guice7/src/main/java/com/fasterxml/jackson/module/guice7/GuiceAnnotationIntrospector.java
Comment (josephlbarnett) [reply guice7/src/main/java/com/fasterxml/jackson/module/guice7/GuiceAnnotationIntrospector.java:1]: oops sorry about that, thanks for fixing!!

Comment (cowtowncoder) [reply guice7/src/main/java/com/fasterxml/jackson/module/guice7/GuiceAnnotationIntrospector.java:1]: np @josephlbarnett ! Just happy I noticed it :)

**Comment by josephlbarnett** [comment]
@cowtowncoder any chance of putting this in a 2.15.3 on a quicker timeline than a 2.16 release?  is there a timeline on 2.16?  (if not, no problem, just curious!)

**Comment by cowtowncoder** [comment]
@josephlbarnett I do not, as a general rule, add new modules in patch releases: it's against SemVer and can be confusing.

No specific timeline for 2.16; I need to send an email since this time around I have specific idea of what needs to be ready for 2.16 -- and as soon as those things are in I'm happy to do RC.
But with vacation time and so on I suspect September is earliest I could so it happening.

**Pull request merged**
